### PR TITLE
DeltaPhi Residual: predict viscous correction over panel-method physics prior

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Instead of predicting absolute flow fields from geometry + conditions, predict only the RESIDUAL between the inviscid panel-method solution and the true CFD solution. This residual is smaller in magnitude, smoother, and focuses the model's capacity on where CFD adds value over potential flow theory — primarily viscous effects (boundary layer separation, wake-induced pressure changes).

DeltaPhi (NeurIPS 2025, Zhejiang Univ) showed 30-40% error reduction in data-limited regimes when predicting residuals from a physics prior. For tandem foils, the dominant viscous effect is wake-induced separation, which the panel method completely misses — making that residual the EXACT target.

We already compute panel-method Cp as an input feature (--cp_panel). This experiment extends that to become the prediction TARGET baseline.

**This is a fundamental change to WHAT the network predicts — not a loss weight or architecture tweak.**

## Instructions

Add a `--residual_from_panel` flag to `train.py`.

### Step 1 — Compute inviscid baseline field

For pressure: The inviscid baseline is already available from the panel-method Cp computation used for `--cp_panel`. Convert panel Cp to dimensional pressure:
```python
p_inviscid = Cp_panel * 0.5 * rho * V^2 + p_ref
```
where rho, V, p_ref come from the flow conditions. If using asinh-transformed pressure, apply the same transform to p_inviscid.

For velocity (Ux, Uy): Use thin-airfoil or panel theory:
```python
# Potential flow around each foil — use the freestream + panel perturbation
Ux_inviscid = V * cos(AoA) + panel_perturbation_x
Uy_inviscid = V * sin(AoA) + panel_perturbation_y
```
If computing panel perturbation velocity is complex, START with pressure residual only (Ux/Uy predicted as absolute, only pressure is residual). This tests the core hypothesis with minimal implementation risk.

### Step 2 — Change prediction targets

In the training loop, change the loss targets:
```python
if residual_from_panel:
    # For pressure: predict dp = p_cfd - p_inviscid
    target_p = p_cfd - p_inviscid  # residual
    # Apply asinh transform to residual if using --asinh_pressure
    
    # For velocity: start with absolute prediction (phase 1)
    target_ux = ux_cfd  # unchanged initially
    target_uy = uy_cfd  # unchanged initially
```

### Step 3 — Reconstruct at inference

At inference time, reconstruct the full field:
```python
if residual_from_panel:
    p_predicted = p_inviscid + model_output_p  # inviscid base + learned residual
```

### Step 4 — Adjust asinh scaling

The residual dp is MUCH smaller in magnitude than absolute p. The current `--asinh_pressure` transform may need re-tuning:
- Check the magnitude of dp = p_cfd - p_inviscid over the training set
- If dp is typically small (|dp| < 0.5 * |p_cfd|), the default asinh parameters should work
- If dp has different scale, add a `--residual_scale` parameter to control the asinh compression

### Step 5 — Training command

```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --anp_srf --residual_from_panel \
  --wandb_group deltaphi-residual \
  --seed 42
```

Run both seeds (42, 73).

### Important: Panel Cp for single-foil vs tandem

Currently `--cp_panel_tandem_only` means panel Cp is zero for single-foil samples. For the residual prediction, you need the panel Cp for ALL samples (both single and tandem), otherwise single-foil samples would have residual = full pressure (defeating the purpose).

**Override `--cp_panel_tandem_only` for this experiment** — compute panel Cp for single-foil samples too (should be straightforward since the panel method works for any airfoil). If this is hard to change, start with tandem-only residual prediction and absolute prediction for single-foil.

### What to Watch

- **p_tan** is the primary target — the tandem wake residual is where the panel method fails most
- **p_in** should improve (attached flow is well-captured by panel, residual is small and smooth)
- Watch for NaN/divergence: if the residual magnitude is very different from the absolute field, gradients may explode. Add gradient clipping if needed.
- If the residual prediction dramatically helps, this is COMBINABLE with Re_x feature, physics constraints, etc.

## Baseline (PR #2379 ANP — current)

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **3.561** | < 3.561 |
| **p_oodc** | **3.847** | < 3.847 |
| **p_tan** | **10.825** | < 10.825 |
| **p_re** | **7.232** | < 7.232 |

W&B runs: metqdxaq (seed 42), jvfvrs4u (seed 73)

Reproduce:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --anp_srf
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

The training command in the PR includes `--anp_srf`, but this flag does not exist in the current codebase on the `noam` branch or this experiment branch. 

Checking git history: the ANP cross-foil decoder code (commit 7999a2e "Add ANP surface decoder") only exists on `frieren/anp-surface-decoder` branch and was never merged into `noam`. The baseline update (fabc3a3) logged ANP results but the code wasn't incorporated.

Should I:
1. Cherry-pick the ANP code from `frieren/anp-surface-decoder` into this branch before implementing DeltaPhi?
2. Implement DeltaPhi on the pre-ANP baseline (without `--anp_srf`)?
3. Wait for the advisor to merge the ANP code into noam first?

The baseline metrics in the PR (p_tan=10.825, p_in=3.561) are from ANP runs, so option 2 would mean comparing against a different (worse) baseline.

**Comment by tcapelle** [comment]
ADVISOR: Good catch, thorfinn. You're right — `--anp_srf` does not exist in the current codebase. The ANP code from PR #2379 was never properly committed to the noam branch (the merge was an empty commit).

**Go with option 2: implement DeltaPhi on the pre-ANP baseline (without `--anp_srf`).** Use the pre-ANP baseline for comparison:

| Metric | Pre-ANP Baseline |
|--------|-----------------|
| p_in | 11.872 |
| p_oodc | 7.459 |
| p_tan | 26.319 |
| p_re | 6.229 |

Your experiment (predicting viscous correction over panel-method prior) is a target reformulation that doesn't depend on ANP — it works with the existing Transolver + SRF architecture. Remove `--anp_srf` from the training command and run immediately.

I'm investigating the ANP code situation separately — once resolved, we can re-run DeltaPhi with ANP. For now, get results on pre-ANP baseline.
